File: src/release-notes/GithubItem.fs
namespace ReleaseNotes

open System.Text
open Octokit

type GitHubItem(issue: Issue, relatedIssues: int list) =  
    member val Issue = issue
    member val RelatedIssues = relatedIssues
    member this.Title =
        let builder = StringBuilder("#")
                          .Append(issue.Number)
                          .Append(" ")       
        if issue.PullRequest = null then
            builder.AppendFormat("[ISSUE] {0}", issue.Title)
        else
            builder.Append(issue.Title) |> ignore
            if relatedIssues.Length > 0 then
                relatedIssues
                |> List.map(fun i -> sprintf "#%i" i)
                |> String.concat ", "
                |> sprintf " (%s: %s)" (if relatedIssues.Length = 1 then "issue" else "issues")
                |> builder.Append
            else builder
        |> ignore                  
        builder.ToString()
        
    member this.TitleAsciiDoc githubUrl =
        if issue.PullRequest = null then
            sprintf "* %s %sissues/%i[#%i]" issue.Title githubUrl issue.Number issue.Number
        else
            let related =
                if relatedIssues.Length > 0 then
                    relatedIssues
                    |> List.map(fun i -> sprintf "%sissues/%i[#%i]" githubUrl i i)
                    |> String.concat ", "
                    |> sprintf " (%s: %s)" (if relatedIssues.Length = 1 then "issue" else "issues")
                else ""
            sprintf "* %s %spull/%i[#%i] %s" issue.Title githubUrl issue.Number issue.Number related
        
    member this.Labels = issue.Labels   
    member this.Number = issue.Number

File: src/release-notes/GitHubItem.cs
using System.Text;
using Octokit;

namespace ReleaseNotes;

public sealed class GitHubItem(Issue issue, IReadOnlyList<int> relatedIssues)
{
	public Issue Issue { get; } = issue;
	public IReadOnlyList<int> RelatedIssues { get; } = relatedIssues;
	public int Number => Issue.Number;
	public IReadOnlyList<Label> Labels => Issue.Labels;

	public string Title
	{
		get
		{
			var builder = new StringBuilder("#").Append(Issue.Number).Append(' ');
			if (Issue.PullRequest is null)
				builder.AppendFormat("[ISSUE] {0}", Issue.Title);
			else
			{
				builder.Append(Issue.Title);
				if (RelatedIssues.Count > 0)
				{
					var related = string.Join(", ", RelatedIssues.Select(i => $"#{i}"));
					var noun = RelatedIssues.Count == 1 ? "issue" : "issues";
					builder.Append($" ({noun}: {related})");
				}
			}

			return builder.ToString();
		}
	}

	public string TitleAsciiDoc(string githubUrl)
	{
		if (Issue.PullRequest is null)
			return $"* {Issue.Title} {githubUrl}issues/{Issue.Number}[#{Issue.Number}]";

		var related = "";
		if (RelatedIssues.Count > 0)
		{
			var links = string.Join(", ", RelatedIssues.Select(i => $"{githubUrl}issues/{i}[#{i}]"));
			var noun = RelatedIssues.Count == 1 ? "issue" : "issues";
			related = $" ({noun}: {links})";
		}

		return $"* {Issue.Title} {githubUrl}pull/{Issue.Number}[#{Issue.Number}] {related}";
	}
}
